### PR TITLE
ON-792 ImaginationMedia_TmwGiftCard: Ignore invoice creation for Bolt order that has a virtual gift card when the hook is sent to Magento

### PR DESCRIPTION
**Pull request by johnmccombs1**
# Description

This PR resolve the following case:
1. The customer placed a Bolt order that has a virtual gift card. The order was created on both Magento and Bolt.
2. The pending hook was sent to Magento and Magento threw the exception “Transaction is not in authorized state.”, 

See the Bugsnag link here:
https://app.bugsnag.com/bolt/magento-2/errors/6102e1040639e600079da9e9?filters%5Bevent.since%5D%5B0%5D%5Btype%5D=eq&filters%5Bevent.since%5D%5B0%5D%5Bvalue%5D=30d&filters%5Bsearch%5D%5B0%5D%5Btype%5D=eq&filters%5Bsearch%5D%5B0%5D%5Bvalue%5D=%22m2dev.shopakira.com%22
This PR resolves the issue by ignoring the invoice creation for Bolt order that has virtual gift card when the hook is sent to Magento

Fixes: https://boltpay.atlassian.net/browse/ON-792

#changelog ON-792 ImaginationMedia_TmwGiftCard: Ignore invoice creation for Bolt order that has a virtual gift card when the hook is sent to Magento

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [ ] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [ ] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# For PR Reviewer 
- [ ] Reviewed unit tests to make sure we are using real components rather than mocks as much as possible?
- [ ] For any major change (observer, new Bolt feature, core Magento interaction) we must add a feature switch, did you verify this?

# Checklist:

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [ ] I have added my Jira ticket link and provided a changelog message.

### CODE REVIEW

File: Plugin/ImaginationMedia/TmwGiftCard/ForceInvoicePlugin.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+     * @param \ImaginationMedia\TmwGiftCard\Observer\ForceInvoice $subject
+     * @param callable $proceed
+     * @param $observer
+     */
+    public function aroundExecute(
+        \ImaginationMedia\TmwGiftCard\Observer\ForceInvoice $subject,
+        callable $proceed,
+        $observer
+    )
+    {
```
Comment (danac-gs) [inline Plugin/ImaginationMedia/TmwGiftCard/ForceInvoicePlugin.php:38]:
```suggestion
    ) {
```

File: Plugin/ImaginationMedia/TmwGiftCard/ForceInvoicePlugin.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+use Bolt\Boltpay\Helper\Hook;
+
+/**
+ * Class ForceInvoicePlugin
+ * @package Bolt\Boltpay\Plugin\ImaginationMedia\TmwGiftCard
+ */
+class ForceInvoicePlugin
+{
+    /**
```
Comment (danac-gs) [inline Plugin/ImaginationMedia/TmwGiftCard/ForceInvoicePlugin.php:28]: Please add a brief description why this is needed.

**Review by danac-gs — CHANGES_REQUESTED** [review]

**Comment by vitaliyreznikov** [comment]
@johnmccombs1 

> Magento threw the exception “Transaction is not in an authorized state.”

Could you please explain why it happens? Not sure I understand.

File: Plugin/ImaginationMedia/TmwGiftCard/ForceInvoicePlugin.php (hunk 2 above)
Comment (johnmccombs1) [reply Plugin/ImaginationMedia/TmwGiftCard/ForceInvoicePlugin.php:28]: @danac-gs  I've added it. Can you check?

**Comment by johnmccombs1** [comment]
@vitaliyreznikov 
When the pending hook is sent to Magento, the $subject observer would try to create an online invoice for the order if it has a virtual gift card and because the Bolt transaction is not yet authorized, an error of “Transaction is not in authorized state.”  would be thrown. We don't need the logic in the $subject observer as our payment/capture hook will generate an offline invoice later. This would bypass the unneeded logic from the $subject observer.

You can see the bugsnag link for the error here: 
https://app.bugsnag.com/bolt/magento-2/errors/6102e1040639e600079da9e9?filters%5Bevent.since%5D%5B0%5D%5Btype%5D=eq&filters%5Bevent.since%5D%5B0%5D%5Bvalue%5D=30d&filters%5Bsearch%5D%5B0%5D%5Btype%5D=eq&filters%5Bsearch%5D%5B0%5D%5Bvalue%5D=%22m2dev.shopakira.com%22

**Comment by danac-gs** [comment]
@johnmccombs1 If I understand correctly, this 3p module triggers invoice creation earlier than usual (expected), while the transaction is still pending. This solution bypasses the online invoice creation at that point and relies on an offline one triggered by out plugin latter. I was wondering how this 3p module triggers the invoice creation and when. Does it happen only once, or, for instance, on every order save (with a check and skip if it was already created). If so, could we skip the invoice creation process while the transaction is pending and let it create it when the transaction is authorized? So a natural, online invoice gets created in Magento.

**Comment by johnmccombs1** [comment]
@danac-gs **Does it happen only once, or, for instance, on every order save (with a check and skip if it was already created)**
=> it happens only once. Their 3p module is using event sales_order_payment_place_start to handle the logic for invoice creation and causing conflicts with Bolt logic.

**If so, could we skip the invoice creation process while the transaction is pending and let it create it when the transaction is authorized? So a natural, online invoice gets created in Magento.**
=> If we do this, we would need to modify a lot of logic in their 3p module which is not an ideal solution in my option.

**Review by danac-gs — APPROVED** [review]

**Pull request merged**